src/main/java/org/elasticsearch/index/translog/fs/BufferingFsTranslogFile.java

```diff
@@ -106,64 +84,51 @@ public long translogSizeInBytes() {
-        assert (((ReentrantReadWriteLock.WriteLock) rwl.writeLock()).isHeldByCurrentThread());
+        assert writeLock.isHeldByCurrentThread();
         if (bufferCount > 0) {
             // we use the channel to write, since on windows, writing to the RAF might not be reflected
             // when reading through the channel
             Channels.writeToChannel(buffer, 0, bufferCount, channelReference.channel());
-
             lastWrittenPosition += bufferCount;
             bufferCount = 0;
         }
     }
 
     @Override
-    public byte[] read(Translog.Location location) throws IOException {
-        rwl.readLock().lock();
-        try {
-            if (location.translogLocation >= lastWrittenPosition) {
-                byte[] data = new byte[location.size];
-                System.arraycopy(buffer, (int) (location.translogLocation - lastWrittenPosition), data, 0, location.size);
-                return data;
+    protected void readBytes(ByteBuffer targetBuffer, long position) throws IOException {
+        try (ReleasableLock _ = readLock.acquire()) {
+            if (position >= lastWrittenPosition) {
+                System.arraycopy(buffer, (int) (position - lastWrittenPosition),
+                        targetBuffer.array(), targetBuffer.position(), targetBuffer.limit());
+                return;
             }
-        } finally {
-            rwl.readLock().unlock();
         }
         // we don't have to have a read lock here because we only write ahead to the file, so all writes has been complete
         // for the requested location.
-        return Channels.readFromFileChannel(channelReference.channel(), location.translogLocation, location.size);
+        Channels.readFromFileChannelWithEofException(channelReference.channel(), position, targetBuffer);
     }
 
-    @Override
-    public FsChannelSnapshot snapshot() throws TranslogException {
-        if (channelReference.tryIncRef()) {
-            boolean success = false;
-            try {
-                rwl.writeLock().lock();
-                try {
-                    flushBuffer();
-                    FsChannelSnapshot snapshot = new FsChannelSnapshot(this.id, channelReference, lastWrittenPosition, operationCounter);
-                    snapshot.seekTo(this.headerSize);
-                    success = true;
-                    return snapshot;
-                } catch (Exception e) {
-                    throw new TranslogException(shardId, "exception while creating snapshot", e);
-                } finally {
-                    rwl.writeLock().unlock();
-                }
-            } finally {
-                if (!success) {
-                    channelReference.decRef();
-                }
+    public FsChannelImmutableReader immutableReader() throws TranslogException {
+        if (channelReference.tryIncRef() == false) {
+            throw new ElasticsearchIllegalStateException("can't increment channel [" + channelReference + "] ref count");
```

Comment:
Can you throw a `TranslogException` here also (encapsulating the `ElasticsearchIllegalStateException` if you want) so that the shardId is captured?